extensions/ql-vscode/src/data-extensions-editor/shared/in-progress-methods.ts

```diff
@@ -1,16 +1,30 @@
 /**
  * A class that keeps track of which methods are in progress for each package.
+ *
+ * This class is immutable and therefore is safe to be used in a react useState hook.
  */
 export class InProgressMethods {
   // A map of in-progress method signatures for each package.
-  private readonly methodMap: Map<string, Set<string>>;
+  private readonly methodMap: ReadonlyMap<string, Set<string>>;
 
-  constructor() {
-    this.methodMap = new Map<string, Set<string>>();
+  constructor(methodMap?: ReadonlyMap<string, Set<string>>) {
+    this.methodMap = methodMap ?? new Map<string, Set<string>>();
   }
 
-  public setPackageMethods(packageName: string, methods: Set<string>): void {
-    this.methodMap.set(packageName, methods);
+  /**
+   * Sets the in-progress methods for the given package.
+   * Returns a new InProgressMethods instance.
+   */
+  public setPackageMethods(
+    packageName: string,
+    methods: Set<string>,
+  ): InProgressMethods {
+    const newMethodMap = new Map<string, Set<string>>();
+    this.methodMap.forEach((value, key) => {
+      newMethodMap.set(key, new Set<string>(value));
+    });
```

Comment:
I believe that these two options should also work and be equivalent:

```suggestion
    const newMethodMap = new Map<string, Set<string>>(this.methodMap);
```

```suggestion
    const newMethodMap = new Map<string, Set<string>>(this.methodMap.entries());
```

These options are probably faster because they can be optimized by V8. I don't think [it's allowed](https://tc39.es/ecma262/multipage/keyed-collections.html#sec-map-constructor) that Node.js uses the same map instance if calling `new Map(this.methodMap)`, so this should be fully equivalent.

Reply:
Thanks for that suggestion. I'd forgotten for this bit that you can pass an iterable to the map constructor.